### ISSUE TITLE
冻结模型

### ISSUE THREAD

**Issue by lenalinyajing** [RRanddom/tf_doc_localisation#4]
你好，我想问一下，冻结模型你是用export_savedmodel来冻结的吗？？如果是，那为什么加载模型并不是用tf.saved_model.loader.load来加载的呢？？如果直接是用ckpt转pb的话，那在测试时graph.get_tensor_by_name(“input：0”）怎么会找不到呢？？望告知

**Comment by yjfncu** [comment]
do you solved this problems?

**Comment by yjfncu** [comment]
inputs = graph.get_tensor_by_name('MobilenetV2/input:0') is ok

**Comment by yjfncu** [comment]
inputs = graph.get_tensor_by_name('MobilenetV2/input:0') is ok

**Comment by weinixuehao** [comment]
![image](https://user-images.githubusercontent.com/17869361/58003426-55777d00-7b13-11e9-9752-99a3776da2cb.png)

![image](https://user-images.githubusercontent.com/17869361/58003674-d898d300-7b13-11e9-98aa-f35285ad9001.png)

checkpoint to pb 需要指定output_node_names吗？或者说需要注意什么事项？
我在代码中没有看到checkpoint to pb 的代码？

**Comment by yjfncu** [comment]
should define output_node_names when converts ckpt model to pb,  you should write convert pb code by yourself

**Comment by weinixuehao** [comment]
@yjfncu 
![image](https://user-images.githubusercontent.com/17869361/58007671-5ca38880-7b1d-11e9-8b99-ba75d8d35c25.png)
Like this? 
Are you chinese？

**Comment by weinixuehao** [comment]
![image](https://user-images.githubusercontent.com/17869361/58008401-e2740380-7b1e-11e9-8379-946dc75a9ebc.png)
The loss reduce so slow? My train images size is croped as 255 X 340.

**Comment by weinixuehao** [comment]
@yjfncu
你有试过test.py 感觉输出结果有问题，绿色框框据左上角。
![image](https://user-images.githubusercontent.com/17869361/58163509-b8554980-7cb6-11e9-8610-767a8918939c.png)
![image](https://user-images.githubusercontent.com/17869361/58163559-ce630a00-7cb6-11e9-956f-ebee13c2a7a3.png)

**Comment by weinixuehao** [comment]
我改成255 X 340 运行test.py能够正常预测了，之前是python版本不对，但是转成tflite格式运行ios还是不行。无法进行预测，对比了下模型结构是一样的。
![image](https://user-images.githubusercontent.com/17869361/58406454-38572700-809c-11e9-8215-ac5b18d18e7a.png)

**Comment by yjfncu** [comment]
@weinixuehao how to transform mode to tflite?

**Comment by weinixuehao** [comment]
@yjfncu 
![image](https://user-images.githubusercontent.com/17869361/58531215-26988f80-8214-11e9-96b7-53efd04a885e.png)
pb文件可以用test.py进行预测，tflite放在手机端无法预测。
![image](https://user-images.githubusercontent.com/17869361/58531264-4def5c80-8214-11e9-9955-4d6a0b75c6e4.png)
这个分辨率降为340X255后测试的结果基于pb文件，但是tflite无法进行预测。

我对比了你的tflite model 和我的tflitemodel input没有区别除了输入size的区别外。
![image](https://user-images.githubusercontent.com/17869361/58531701-dde1d600-8215-11e9-85a6-7b16cdcec942.png)

**Comment by weinixuehao** [comment]
![image](https://user-images.githubusercontent.com/17869361/58533672-b5f57100-821b-11e9-9783-d99a0e4dca84.png)
假如不做dsnt之前heatmap 对比效果
上面的是你的模型600X800 
下面是我的模型255X340

**Comment by yjfncu** [comment]
@weinixuehao 你微信多少

**Comment by weinixuehao** [comment]
@yjfncu 微信:13755183280 qq:253861581

**Comment by weinixuehao** [comment]
@yjfncu 加我了没

**Comment by LanWong1** [comment]
![image](https://user-images.githubusercontent.com/33218397/61696028-21187b00-ad67-11e9-85f6-6dbd22332120.png)
请问为什么我的模型输入是个 iterator？而不是简单的 input？

**Comment by Antonio-hi** [comment]
`# -*- coding: utf-8 -*-
# @Author        : lk
# @Email         : 9428.al@gmail.com
# @Created Date  : 2019-08-16 11:26:50
# @Last Modified : 2019-08-16 14:58:29
# @Description   : .ckpt转化为pb模型

import os 
import numpy as np 
import tensorflow as tf
from model.keypoints_heatmaps_model import main_network

tf.app.flags.DEFINE_string('checkpoint_path', './data/train_logdir/', '')
tf.app.flags.DEFINE_string('pb_model_path', './frozen_model/', '')
tf.app.flags.DEFINE_integer('image_height', 800, '')
tf.app.flags.DEFINE_integer('image_width', 600, '')

FLAGS = tf.app.flags.FLAGS

w, h = FLAGS.image_height, FLAGS.image_width

def freeze_graph():
    input_images = tf.placeholder(tf.float32, shape=[None, w, h, 3], name='input')
    keypoints_pred, heatmaps_pred, keypoints, heatmaps  = main_network(input_images)

    with tf.Session(config=tf.ConfigProto(allow_soft_placement=True)) as sess:
        ckpt = tf.train.latest_checkpoint(FLAGS.checkpoint_path)
        print('[INFO] restore previous checkpoint from: ', ckpt)
        saver = tf.train.Saver()
        saver.restore(sess, ckpt)

        output_graph_def = tf.graph_util.convert_variables_to_constants(sess,
                            input_graph_def = sess.graph.as_graph_def(),
                            output_node_names =  ["keypoints_pred", "heatmaps"])
        names = [n.name for n in output_graph_def.node]
        for name in names:
            print (name)

    output_graph = "frozen_model_new.pb"
    output_path = os.path.join(FLAGS.pb_model_path, output_graph)
    with tf.gfile.GFile(output_path, "wb") as f:
        f.write(output_graph_def.SerializeToString())
    print("%d ops in the final graph" % len(output_graph_def.node))

    return None

if __name__ == "__main__":
    freeze_graph()`

**Comment by jiangzz1628** [comment]
> @yjfncu 喜欢这个吗？ 你是中国人吗？
> ![图片](https://user-images.githubusercontent.com/17869361/58007671-5ca38880-7b1d-11e9-8b99-ba75d8d35c25.png)

您好，[n.name for n in tf.get_default_graph().as_graph_def().node] 得到的结果是[],这个节点您知道叫什么名字吗

**Comment by jiangzz1628** [comment]
> 你好，我想问一下，冻结模型你是用export_savedmodel来冻结的吗？？？是，那为什么加载模型本身用tf.saved_model.loader.load来加载的呢？？如果直接是用ckpt转pb的话，那在测试时graph.get_tensor_by_name（“输入：0”）怎么会找到呢？？望告知

您好，您用ckpt直接转的pb吗？那个节点的名字叫什么，还请告知一下，谢谢

**Comment by clevertension** [comment]
![image](https://user-images.githubusercontent.com/346723/105661065-7955ad80-5f07-11eb-8d8b-0916245632d7.png)

slim.batch_norm的时候， 一个mean和variance没有更新，一个有更新，这个是什么原因呢?

**Comment by maokp** [comment]
@weinixuehao pb模型转换为tflite后结果不对，请问这个问题解决了吗？

**Comment by JaosonMa** [comment]
> 什么原因呢

请问你解决这个问题了吗？我导出的也有问题没更新

**Comment by JaosonMa** [comment]
> 

你能提供一下，如何转.pb的代码吗？ 我现在不管怎么转都跟作者提供的有区别，mean,variance没有参数，然后导出模型最后也不一样
